backend/cfg/cfg.ml

```diff
@@ -97,24 +83,34 @@ module BasicInstructionList = struct
     then cell.t.last <- new_node
     else new_node.next.prev <- new_node
 
-  let instr cell = cell.node.instr
+  let value cell =
+    assert (not (is_dummy_node cell.node));
+    cell.node.value
 
-  let make_empty () = { length = 0; first = dummy_node; last = dummy_node }
+  let prev cell =
+    assert (not (is_dummy_node cell.node));
+    let prev = cell.node.prev in
+    match is_dummy_node prev with
+    | true -> None
+    | false -> Some { node = prev; t = cell.t }
 
-  let make_single instr =
-    let node = unattached_node instr in
+  let make_empty () =
+    { length = 0; first = get_dummy_node (); last = get_dummy_node () }
+
+  let make_single value =
+    let node = unattached_node value in
     { length = 1; first = node; last = node }
 
   let hd t =
     let first = t.first in
-    if is_dummy_node first then None else Some first.instr
+    if is_dummy_node first then None else Some first.value
 
   let last t =
     let last = t.last in
-    if is_dummy_node last then None else Some last.instr
+    if is_dummy_node last then None else Some last.value
 
-  let add_begin t instr =
-    let node = unattached_node instr in
+  let add_begin t value =
+    let node = unattached_node value in
     let len = t.length in
     if Int.equal len 0
```

Comment:
in that case, I'd add some asserts to make sure that first and last are dummy nodes